File: leetcode_cpp/invert-binary-tree.cpp
#include <iostream>
#include <cassert>

#include <vector>
#include <stack>

#include "../tree.h"

using namespace std;

/**
https://leetcode.com/problems/invert-binary-tree/

// Solution1. Recursion. DSF inorder
// time: o(n)
// space: o(n)
node* recursion(node*)
    if (!node) return nullptr
    auto l = recursion(node->left);
    auto r = recursion(node->right);
    node->left = r;
    node->right = l;
    return node;
    
// Solution2. Iteration BSF with queue
// time: o(n)
// space: o(n)
BSF
q: (4)
swap 4-left, 4-right
q: (2) 7
swap 2-left, 2-right
q: (7) 1 3 6 9


*/
class Solution {
public:
    TreeNode* invertTree_recursion(TreeNode* root) {
        if (!root) return nullptr;
        auto l = invertTree_recursion(root->left);
        auto r = invertTree_recursion(root->right);
        root->left = r;
        root->right = l;
        return root;
    }
    
    TreeNode* invertTree_iteration(TreeNode* root) {
        if (!root) return nullptr;
        queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            auto node = q.front();
            q.pop();
            auto t = node->left;
            node->left = node->right;
            node->right = t;
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
        
        return root;
    }
    
    TreeNode* invertTree(TreeNode* root) {
        //return invertTree_recursion(root);
        return invertTree_iteration(root);
    }
};

int main()
{
    auto input1 = new TreeNode(4);
    input1->left = new TreeNode(2);
    input1->left->left = new TreeNode(1);
    input1->left->right = new TreeNode(3);
    input1->right = new TreeNode(7);
    input1->right->left = new TreeNode(6);
    input1->right->right = new TreeNode(9);
    auto ans = Solution().invertTree(input1);
    assert(ans->left->val == 7 && ans->left->left->val == 9);

    return 0;
}

File: leetcode/leetcode_py/invert-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

'''
 https://leetcode.com/problems/invert-binary-tree/
 
 Solution1. recursion. DSF
 time: o(n)
 space: o(n)
 
 Solution2. iteration. BSF
 time: o(n)
 space: o(n)
 '''
 
class Solution:
    def invertTree_1(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root: return None
        l = self.invertTree_1(root.left)
        r = self.invertTree_1(root.right)
        root.left = r
        root.right = l
        return root

    def invertTree_2(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root: return None
        q = deque()
        q.append(root)        
        while len(q):
            n = len(q)
            for i in range(n):
                node = q.popleft()
                
                if node.left: q.append(node.left)
                if node.right: q.append(node.right)

                temp = node.left
                node.left = node.right
                node.right = temp              

        return root

    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        return self.invertTree_2(root)